library/std/src/io/buffered/bufwriter.rs

```diff
@@ -287,47 +301,115 @@ impl<W: Write> BufWriter<W> {
             r
         } else {
-            self.buf.extend_from_slice(buf);
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
+            // Therefore, `self.buf.len() + buf.len() <= self.buf.capacity()`.
+            unsafe {
+                self.write_to_buffer_unchecked(buf);
+            }
+
             Ok(buf.len())
         }
     }
 
-    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+    // Ensure this function does not get inlined into `write_all`, so that it
+    // remains inlineable and its common path remains as short as possible.
+    // If this function ends up being called frequently relative to `write_all`,
+    // it's likely a sign that the client is using an improperly sized buffer.
+    #[inline(never)]
+    fn flush_and_write_all(&mut self, buf: &[u8]) -> io::Result<()> {
         // Normally, `write_all` just calls `write` in a loop. We can do better
         // by calling `self.get_mut().write_all()` directly, which avoids
         // round trips through the buffer in the event of a series of partial
         // writes in some circumstances.
-        if self.buf.len() + buf.len() > self.buf.capacity() {
-            self.flush_buf()?;
-        }
+        self.flush_buf()?;
+
         // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
         if buf.len() >= self.buf.capacity() {
             self.panicked = true;
             let r = self.get_mut().write_all(buf);
             self.panicked = false;
             r
         } else {
-            self.buf.extend_from_slice(buf);
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
+            // Therefore, `self.buf.len() + buf.len() <= self.buf.capacity()`.
+            unsafe {
+                self.write_to_buffer_unchecked(buf);
+            }
+
             Ok(())
         }
     }
 
+    // SAFETY: Requires `self.buf.len() + buf.len() <= self.buf.capacity()`,
+    // i.e., that input buffer length is less than or equal to spare capacity.
+    #[inline(always)]
+    unsafe fn write_to_buffer_unchecked(&mut self, buf: &[u8]) {
+        debug_assert!(self.buf.len() + buf.len() <= self.buf.capacity());
+        let old_len = self.buf.len();
+        let buf_len = buf.len();
+        let src = buf.as_ptr();
+        let dst = self.buf.as_mut_ptr().add(old_len);
+        ptr::copy_nonoverlapping(src, dst, buf_len);
+        self.buf.set_len(old_len + buf_len);
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<W: Write> Write for BufWriter<W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Write to the buffer even if it exactly fills it. Trying to avoid this
+        // adds cost to the overwhelmingly common case where it doesn't happen.
```

Comment:
That is not quite true. `io::copy` uses a `DEFAULT_BUF_SIZE`-sized buffer internally. `BufWriter::new`  also sets its capacity to `DEFAULT_BUF_SIZE`, which means if you toss a `BufWriter` into `io::copy` then for every single iteration of the copy loop it will try to write exactly as much as needed to fill that buffer, i.e. doubles the `memcpy` traffic.

That said, this can be solved with specialization and #78641 does just that.

Reply:
Okay, I can roll that change back. It's the least effective of the optimizations here, and rolling it back may avoid other pathological cases.

Reply:
Rolled it back. Thanks @the8472. If you think that this change should be reimplemented once #78641 is done, please feel free to add it.

Reply:
I managed to reimplement the optimization in a way that is a win for what is often the common case, while not hurting the case you mentioned or pathological cases much. Latest commit message gives the justification.